### PR TITLE
kafka(ticdc): reset the admin client to fix broken pipe

### LINKED ISSUE #8223: Kakfa changefeed stucks
### What did you do?

1. create kafka changefeed with big table scale out is enabled
```
[scheduler] 
region-per-span = 1_000  
```
2. Use lightning to import ~300GB data to upstream via TiDB backend.
3. Check changefeed status

### What did you expect to see?

Changefeed should adavance normally

### What did you see instead?

Changefeed stucks

687ea2f3-d291-4fa8-9617-514fa9e074c5

### Versions of the cluster

Upstream TiDB cluster version (execute `SELECT tidb_version();` in a MySQL client):

```console
[root@upstream-tidb-0 /]# /tidb-server -V                
Release Version: v6.6.0-alpha                            
Edition: Community                                       
Git Commit Hash: 93c6492418ae1bfa91479b8b465e61c78f29a995
Git Branch: heads/refs/tags/v6.6.0-alpha                 
UTC Build Time: 2023-02-08 11:51:47                      
GoVersion: go1.19.5                                      
Race Enabled: false                                      
TiKV Min Version: 6.2.0-alpha                            
Check Table Before Drop: false                           
Store: unistore                                          
```

Upstream TiKV version (execute `tikv-server --version`):

```console
[root@upstream-tikv-0 /]# /tikv-server -V                                                                                                  
TiKV                                                                                                                                       
Release Version:   6.6.0-alpha                                                                                                             
Edition:           Community                                                                                                               
Git Commit Hash:   d513b789f0709520cdccc20bc4e6560761ad01a5                                                                                
Git Commit Branch: heads/refs/tags/v6.6.0-alpha                                                                                            
UTC Build Time:    2023-02-09 11:50:03                                                                                                     
Rust Version:      rustc 1.67.0-nightly (96ddd32c4 2022-11-14)                                                                             
Enable Features:   pprof-fp jemalloc mem-profiling portable sse test-engine-kv-rocksdb test-engine-raft-raft-engine cloud-aws cloud-gcp clo
ud-azure                                                                                                                                   
Profile:           dist_release                                                                                                            
```

TiCDC version (execute `cdc version`):

```console
[2023/02/09 10:46:54.599 +00:00] [INFO] [version.go:47] ["Welcome to Change Data Capture (CDC)"] [release-version=v6.6.0-alpha] [git-hash=0e1322f90b9b72697ca74513c32c0cb64479064c] [git-branch=heads/refs/tags/v6.6.0-alpha] [utc-build-time="2023-02-08 11:41:41"] [go-version="go version go1.19.5 linux/amd64"] [failpoint-build=false]                                                                                                                                                                   
```

### LINKED ISSUE #8225: kafka sink not robust to admin tasks
Kafka admin task meet error, cause the whole changefeed restart. 

```
[2023/02/09 15:42:53.245 +00:00] [ERROR] [manager.go:199] ["Worker handles sink task failed"] [namespace=default] [changefeed=kafka-scale-big-table] [error="write tcp 10.233.84.234:54958->10.233.122.44:9092: write: broken pipe"] [errorVerbose="write tcp 10.233.84.234:54958->10.233.122.44:9092: write: broken pipe\ngithub.com/pingcap/errors.AddStack\n\tgithub.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/errors.go:174\ngithub.com/pingcap/errors.Trace\n\tgithub.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/juju_adaptor.go:15\ngithub.com/pingcap/tiflow/cdc/sink/mq/manager.(*kafkaTopicManager).GetPartitionNum\n\tgithub.com/pingcap/tiflow/cdc/sink/mq/manager/kafka_manager.go:70\ngithub.com/pingcap/tiflow/cdc/sinkv2/eventsink/mq.(*dmlSink).WriteEvents\n\tgithub.com/pingcap/tiflow/cdc/sinkv2/eventsink/mq/mq_dml_sink.go:120\ngithub.com/pingcap/tiflow/cdc/sinkv2/tablesink.(*EventTableSink[...]).UpdateResolvedTs\n\tgithub.com/pingcap/tiflow/cdc/sinkv2/tablesink/table_sink_impl.go:115\ngithub.com/pingcap/tiflow/cdc/processor/sinkmanager.(*tableSinkWrapper).updateResolvedTs\n\tgithub.com/pingcap/tiflow/cdc/processor/sinkmanager/table_sink_wrapper.go:170\ngithub.com/pingcap/tiflow/cdc/processor/sinkmanager.(*sinkWorker).advanceTableSink\n\tgithub.com/pingcap/tiflow/cdc/processor/sinkmanager/table_sink_worker.go:459\ngithub.com/pingcap/tiflow/cdc/processor/sinkmanager.(*sinkWorker).handleTask.func2\n\tgithub.com/pingcap/tiflow/cdc/processor/sinkmanager/table_sink_worker.go:167\ngithub.com/pingcap/tiflow/cdc/processor/sinkmanager.(*sinkWorker).handleTask.func3\n\tgithub.com/pingcap/tiflow/cdc/processor/sinkmanager/table_sink_worker.go:214\ngithub.com/pingcap/tiflow/cdc/processor/sinkmanager.(*sinkWorker).handleTask\n\tgithub.com/pingcap/tiflow/cdc/processor/sinkmanager/table_sink_worker.go:320\ngithub.com/pingcap/tiflow/cdc/processor/sinkmanager.(*sinkWorker).handleTasks\n\tgithub.com/pingcap/tiflow/cdc/processor/sinkmanager/table_sink_worker.go:76\ngithub.com/pingcap/tiflow/cdc/processor/sinkmanager.(*SinkManager).startWorkers.func1\n\tgithub.com/pingcap/tiflow/cdc/processor/sinkmanager/manager.go:197\nruntime.goexit\n\truntime/asm_amd64.s:1594"]
[2023/02/09 15:42:53.252 +00:00] [ERROR] [processor.go:745] ["error on running processor"] [capture=47d9ab45-1bae-4061-96fd-d0f2763e411d] [namespace=default] [changefeed=kafka-scale-big-table] [error="write tcp 10.233.84.234:54958->10.233.122.44:9092: write: broken pipe"] [errorVerbose="write tcp 10.233.84.234:54958->10.233.122.44:9092: write: broken pipe\ngithub.com/pingcap/errors.AddStack\n\tgithub.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/errors.go:174\ngithub.com/pingcap/errors.Trace\n\tgithub.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/juju_adaptor.go:15\ngithub.com/pingcap/tiflow/cdc/sink/mq/manager.(*kafkaTopicManager).GetPartitionNum\n\tgithub.com/pingcap/tiflow/cdc/sink/mq/manager/kafka_manager.go:70\ngithub.com/pingcap/tiflow/cdc/sinkv2/eventsink/mq.(*dmlSink).WriteEvents\n\tgithub.com/pingcap/tiflow/cdc/sinkv2/eventsink/mq/mq_dml_sink.go:120\ngithub.com/pingcap/tiflow/cdc/sinkv2/tablesink.(*EventTableSink[...]).UpdateResolvedTs\n\tgithub.com/pingcap/tiflow/cdc/sinkv2/tablesink/table_sink_impl.go:115\ngithub.com/pingcap/tiflow/cdc/processor/sinkmanager.(*tableSinkWrapper).updateResolvedTs\n\tgithub.com/pingcap/tiflow/cdc/processor/sinkmanager/table_sink_wrapper.go:170\ngithub.com/pingcap/tiflow/cdc/processor/sinkmanager.(*sinkWorker).advanceTableSink\n\tgithub.com/pingcap/tiflow/cdc/processor/sinkmanager/table_sink_worker.go:459\ngithub.com/pingcap/tiflow/cdc/processor/sinkmanager.(*sinkWorker).handleTask.func2\n\tgithub.com/pingcap/tiflow/cdc/processor/sinkmanager/table_sink_worker.go:167\ngithub.com/pingcap/tiflow/cdc/processor/sinkmanager.(*sinkWorker).handleTask.func3\n\tgithub.com/pingcap/tiflow/cdc/processor/sinkmanager/table_sink_worker.go:214\ngithub.com/pingcap/tiflow/cdc/processor/sinkmanager.(*sinkWorker).handleTask\n\tgithub.com/pingcap/tiflow/cdc/processor/sinkmanager/table_sink_worker.go:320\ngithub.com/pingcap/tiflow/cdc/processor/sinkmanager.(*sinkWorker).handleTasks\n\tgithub.com/pingcap/tiflow/cdc/processor/sinkmanager/table_sink_worker.go:76\ngithub.com/pingcap/tiflow/cdc/processor/sinkmanager.(*SinkManager).startWorkers.func1\n\tgithub.com/pingcap/tiflow/cdc/processor/sinkmanager/manager.go:197\nruntime.goexit\n\truntime/asm_amd64.s:1594"]
[2023/02/09 15:42:53.252 +00:00] [ERROR] [processor.go:539] ["run processor failed"] [capture=47d9ab45-1bae-4061-96fd-d0f2763e411d] [namespace=default] [changefeed=kafka-scale-big-table] [error="write tcp 10.233.84.234:54958->10.233.122.44:9092: write: broken pipe"] [errorVerbose="write tcp 10.233.84.234:54958->10.233.122.44:9092: write: broken pipe\ngithub.com/pingcap/errors.AddStack\n\tgithub.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/errors.go:174\ngithub.com/pingcap/errors.Trace\n\tgithub.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/juju_adaptor.go:15\ngithub.com/pingcap/tiflow/cdc/sink/mq/manager.(*kafkaTopicManager).GetPartitionNum\n\tgithub.com/pingcap/tiflow/cdc/sink/mq/manager/kafka_manager.go:70\ngithub.com/pingcap/tiflow/cdc/sinkv2/eventsink/mq.(*dmlSink).WriteEvents\n\tgithub.com/pingcap/tiflow/cdc/sinkv2/eventsink/mq/mq_dml_sink.go:120\ngithub.com/pingcap/tiflow/cdc/sinkv2/tablesink.(*EventTableSink[...]).UpdateResolvedTs\n\tgithub.com/pingcap/tiflow/cdc/sinkv2/tablesink/table_sink_impl.go:115\ngithub.com/pingcap/tiflow/cdc/processor/sinkmanager.(*tableSinkWrapper).updateResolvedTs\n\tgithub.com/pingcap/tiflow/cdc/processor/sinkmanager/table_sink_wrapper.go:170\ngithub.com/pingcap/tiflow/cdc/processor/sinkmanager.(*sinkWorker).advanceTableSink\n\tgithub.com/pingcap/tiflow/cdc/processor/sinkmanager/table_sink_worker.go:459\ngithub.com/pingcap/tiflow/cdc/processor/sinkmanager.(*sinkWorker).handleTask.func2\n\tgithub.com/pingcap/tiflow/cdc/processor/sinkmanager/table_sink_worker.go:167\ngithub.com/pingcap/tiflow/cdc/processor/sinkmanager.(*sinkWorker).handleTask.func3\n\tgithub.com/pingcap/tiflow/cdc/processor/sinkmanager/table_sink_worker.go:214\ngithub.com/pingcap/tiflow/cdc/processor/sinkmanager.(*sinkWorker).handleTask\n\tgithub.com/pingcap/tiflow/cdc/processor/sinkmanager/table_sink_worker.go:320\ngithub.com/pingcap/tiflow/cdc/processor/sinkmanager.(*sinkWorker).handleTasks\n\tgithub.com/pingcap/tiflow/cdc/processor/sinkmanager/table_sink_worker.go:76\ngithub.com/pingcap/tiflow/cdc/processor/sinkmanager.(*SinkManager).startWorkers.func1\n\tgithub.com/pingcap/tiflow/cdc/processor/sinkmanager/manager.go:197\nruntime.goexit\n\truntime/asm_amd64.s:1594"]
```

### PR DESCRIPTION
**Pull request by 3AceShowHand**
### What problem does this PR solve?

Issue Number: close #8225, close #8223

### What is changed and how it works?

`AdminClient` may meet `write: broken pipe`, it's a know issue have not been addressed yet.

* reset the `admin client` if meet `broken pipe` error.

#### Tests 

 - Manual test (add detailed scripts or steps below)

### Release note 

```release-note
None`
```

### CODE REVIEW

File: pkg/sink/kafka/admin.go [hunk 1]
```diff
@@ -39,11 +47,44 @@ func NewSaramaAdminClient(ctx context.Context, config *Options) (ClusterAdminCli
...
+		brokers, _, err = a.client.DescribeCluster()
+		if err == nil {
+			return nil
+		}
+
+		if !errors.Is(err, syscall.EPIPE) {
+			return err
+		}
+
+		return a.reset()
```
Comment (amyangfei) [inline pkg/sink/kafka/admin.go:86]: if `a.reset` returns nil, retry will not run and brokers won't be refreshed.

File: pkg/sink/kafka/admin.go [hunk 2]
```diff
@@ -39,11 +47,44 @@ func NewSaramaAdminClient(ctx context.Context, config *Options) (ClusterAdminCli
...
+	}, retry.WithBackoffBaseDelay(20), retry.WithMaxTries(1))
```
Comment (amyangfei) [inline pkg/sink/kafka/admin.go:87]: `retry.WithMaxTries(1)` will run func in `retry.Do()` only once.

File: pkg/sink/kafka/admin.go [hunk 3]
```diff
@@ -90,11 +162,24 @@ func (a *saramaAdminClient) GetBrokerConfig(_ context.Context, configName string
...
+		defer a.mu.Unlock()
+		topics, err = a.client.ListTopics()
+		if err == nil {
+			return nil
+		}
+		if !errors.Is(err, syscall.EPIPE) {
+			return err
+		}
+		return a.reset()
+	}, retry.WithBackoffBaseDelay(20), retry.WithMaxTries(1))
```
Comment (amyangfei) [inline pkg/sink/kafka/admin.go:183]: `err` returned in L171 is not checked.

File: pkg/sink/kafka/admin.go [hunk 4]
```diff
@@ -39,11 +52,44 @@ func NewSaramaAdminClient(ctx context.Context, config *Options) (ClusterAdminCli
...
+func (a *saramaAdminClient) reset() error {
+	newClient, err := sarama.NewClusterAdmin(a.brokerEndpoints, a.config)
+	if err != nil {
+		return cerror.Trace(err)
+	}
+
+	_ = a.client.Close()
+	a.client = newClient
+
+	return errors.New("retry after reset")
```
Comment (sdojjy) [inline pkg/sink/kafka/admin.go:71]: why do we always return an error?

File: pkg/sink/kafka/admin.go [hunk 5]
```diff
@@ -39,11 +52,44 @@ func NewSaramaAdminClient(ctx context.Context, config *Options) (ClusterAdminCli
...
 }
 
-func (a *saramaAdminClient) GetAllBrokers(context.Context) ([]Broker, error) {
-	brokers, _, err := a.client.DescribeCluster()
+func (a *saramaAdminClient) GetAllBrokers(ctx context.Context) ([]Broker, error) {
+	var (
+		brokers []*sarama.Broker
+		err     error
+	)
+	err = retry.Do(ctx, func() error {
```
Comment (sdojjy) [inline pkg/sink/kafka/admin.go:79]: all the retry logic is almost the same, can you refine it to reuse same code?

Comment (zhaoxinyu) [reply pkg/sink/kafka/admin.go:79]: +1, Could the retry logic be extracted to a common place?

**Comment by zhaoxinyu** [comment]
I'm wondering what is the route cause of this "broken pipe" error? Is the original issue related to this sarama issue (https://github.com/Shopify/sarama/issues/2173)? Maybe it's a sarama bug which hasn't been fixed thoroughly.

**Comment by 3AceShowHand** [comment]
> I'm wondering what is the route cause of this "broken pipe" error? Is the original issue related to this sarama issue ([Shopify/sarama#2173](https://github.com/Shopify/sarama/issues/2173))? Maybe it's a sarama bug which hasn't been fixed thoroughly.

yes, it's a know bug issue last for a few years, which has not been addressed yet.

File: pkg/sink/kafka/admin.go (hunk 4 above)
Comment (3AceShowHand) [reply pkg/sink/kafka/admin.go:71]: if return nil here, the retry logic will exit, this return an error is on purpose.

File: pkg/sink/kafka/admin.go (hunk 5 above)
Comment (3AceShowHand) [reply pkg/sink/kafka/admin.go:79]: yes, we can do it, but let us do it when it's not in a hurry.

Comment (3AceShowHand) [reply pkg/sink/kafka/admin.go]: done, we can use high order function to solve this.

File: pkg/sink/kafka/admin.go (hunk 4 above)
Comment (CharlesCheung96) [reply pkg/sink/kafka/admin.go:72]: Perhaps it is more appropriate to return this error outside of reset?

Comment (sdojjy) [reply pkg/sink/kafka/admin.go:72]: it's very strange that a func always returns an error.

**Review by sdojjy — APPROVED** [review]

File: pkg/sink/kafka/admin.go (hunk 4 above)
Comment (3AceShowHand) [reply pkg/sink/kafka/admin.go:72]: if we move it outside, we still have to write one error when retry, no matter reset return error or not.

**Review by CharlesCheung96 — APPROVED** [review]

**Pull request merged**